server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java

```diff
@@ -80,9 +79,47 @@
      * @throws MessagingException when the message is not clonable
      */
     public static MailImpl duplicate(Mail mail) throws MessagingException {
-        return new MailImpl(mail, newName(mail));
+        return new MailImpl(mail, deriveNewName(mail.getName()));
     }
 
+    private static final java.util.Random random = new java.util.Random(); // Used
+    // to
+    // generate
+    // new
+    // mail
+    // names
+
+    /**
+     * Create a unique new primary key name for the given MailObject.
+     *
+     * @param currentName the mail to use as the basis for the new mail name
+     * @return a new name
+     */
+    @VisibleForTesting static String deriveNewName(String currentName) throws MessagingException {
+
+        // Checking if the original mail name is too long, perhaps because of a
+        // loop caused by a configuration error.
+        // it could cause a "null pointer exception" in AvalonMailRepository
+        // much
+        // harder to understand.
+        if (currentName.length() > 76) {
+            int count = 0;
```

Comment:
int count = StringUtils.countMatches(currentName, '!');

And drop index.

If you are moving code, you should refactor it ;-)

Reply:
please read the whole review before saying such things : https://github.com/linagora/james-project/pull/1126/commits/52c2b3d7dd9f8b4b25c5a70fb6b9f25cda82cef6